zerver/lib/message.py

```diff
@@ -1095,22 +1092,15 @@ def get_recent_private_conversations(user_profile: UserProfile) -> Dict[int, Dic
         LIMIT %(conversation_limit)d)
         UNION ALL
         (SELECT
-            um.message_id AS message_id,
-            r.id AS recipient_id
+            m.id AS message_id,
+            sender_profile.personal_recipient_id AS recipient_id
         FROM
-            zerver_usermessage um
-        JOIN
             zerver_message m
-        ON
-            um.message_id = m.id
         JOIN
-            zerver_recipient r
+            zerver_userprofile sender_profile
         ON
-            r.type = 1 AND
-            r.type_id = m.sender_id
+            m.sender_id = sender_profile.id
         WHERE
-            um.user_profile_id=%(user_profile_id)d AND
-            um.flags & 2048 <> 0 AND
```

Comment:
I think we can't remove the `UserMessage` join because we need to confirm it's unread, don't we?

Reply:
Hmm, I don't see where we were checking for 'unread' though? `` um.flags & 2048 <> 0`` checks the ``is_private`` bitflag I believe.

Though thinking about this function more, what it does right now is return up to ``2*RECENT_CONVERSATIONS_LIMIT`` latest conversations (latest in the sense of highest message_id in the conversation). ``2*`` because we're ``UNION``-ing two queries limited by ``RECENT_CONVERSATIONS_LIMIT``. So since we're just going by "highest message id" here, whether a message was read or not doesn't come into play.

But this leaves a perhaps undesirable rare case, where the user has a still unread message in a very ancient conversation. If there's sufficiently many newer conversations, the conversation with the unread message won't be returned here. Perhaps it should? On the other hand, this is a somewhat absurd scenario requiring *at least* 1000 new private conversation after leaving a message unread for this issue to play out.

Reply:
Oops, I was confused!  I think your implementation here is correct; I was misreading the code.

Unread messages are handled by a separate query for all unread messages using that index; this data set is intended to be used to sort "recent private message conversations".  So I think we explicitly don't care about the unread flag here.